src/Raku/ast/base.rakumod

```diff
@@ -217,28 +217,68 @@ class RakuAST::Node {
         }
     }
 
-    method dump(int $indent?) {
-        my str $prefix := nqp::x(' ', $indent);
-        my $name := nqp::substr(self.HOW.name(self), nqp::chars('RakuAST::'));
-
+    method dump-markers() {
         my @markers;
         @markers.push('⚓') if nqp::istype(self, RakuAST::Sinkable) && self.sunk;
         @markers.push('▪') if nqp::istype(self, RakuAST::BlockStatementSensitive) && self.is-block-statement;
-        my $markers := @markers ?? ' ' ~ nqp::join('', @markers) !! '';
-
-        my $dump := "$prefix$name$markers\n";
-        if nqp::istype(self, RakuAST::Expression) {
-            self.visit-thunks(-> $thunk {
-                $dump := "$dump$prefix  🧠 " ~ $thunk.thunk-kind ~ "\n";
-                $thunk.visit-children(-> $child {
-                    $dump := $dump ~ $child.dump($indent + 4);
-                });
-            });
+        if nqp::isconcrete($!origin) {
+            @markers.push('𝄞') if $!origin.is-key();
         }
+        nqp::join('', @markers)
+    }
+
+    method dump-extras(int $indent) { '' }
```

Comment:
Shouldn’t this be something like ‘nqp::x(“ “, $indent)’?

Reply:
You mean like get the indent as a prepared string? I decided that integer provides more flexibility for custom methods.

Reply:
I mean it isn't using the `$indent` parameter at all.

Reply:
Extras may happen to be a multi-liner. In this case either option must be implemented by the caller:

- require a positional return value
- split on new lines

Not sure if either is better than passing the indent value. Though positional return doesn't look that terrible. Perhaps I would consider this variant.

Reply:
Maybe I'm just not seeing the overall picture, but the comment seems to contradict the implementation? "... expected to respect the indenation level provided with $indent", but the implementation doesn't respect (in fact it completely ignores) $indent. Can you show an example of how it would look?
Unrelated, but "indenation" -> "indentation"

Reply:
This one is a stub and produces no output. Look into _expressions.rakumod_. This is where it does some useful work.

The comment describes requirements to more material implementations of the method.